File: rust/data-structures/src/arrays/ex1_find_max_element.rs
/// Exercise: Find Maximum Element (Easy)
///
/// Given a slice of integers, find and return the maximum element.
///
/// Example:
/// Input: &[3, 7, 2, 9, 1]
/// Output: Some(9)
///
/// Time Complexity: O(n)
/// Space Complexity: O(1)

pub fn find_max(arr: &[i32]) -> Option<i32> {
    // TODO: Implement this function
    // Edge case: handle empty slice (return None)
    // Hint: Use iter().max() or manual iteration
    // Remember to dereference when using iter().max()
    unimplemented!("Not implemented yet")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_basic_case() {
        assert_eq!(find_max(&[3, 7, 2, 9, 1]), Some(9));
    }

    #[test]
    fn test_single_element() {
        assert_eq!(find_max(&[42]), Some(42));
    }

    #[test]
    fn test_negative_numbers() {
        assert_eq!(find_max(&[-5, -2, -8, -1, -10]), Some(-1));
    }

    #[test]
    fn test_all_same_elements() {
        assert_eq!(find_max(&[5, 5, 5, 5]), Some(5));
    }

    #[test]
    fn test_max_at_beginning() {
        assert_eq!(find_max(&[10, 5, 3, 2, 1]), Some(10));
    }

    #[test]
    fn test_max_at_end() {
        assert_eq!(find_max(&[1, 2, 3, 4, 15]), Some(15));
    }

    #[test]
    fn test_empty_slice() {
        assert_eq!(find_max(&[]), None);
    }
}

File: java/src/main/java/com/learning/datastructures/arrays/ex1_FindMaxElement.java
package com.learning.datastructures.arrays;

/**
 * Exercise: Find Maximum Element (Easy)
 *
 * Given an array of integers, find and return the maximum element.
 *
 * Example:
 * Input: [3, 7, 2, 9, 1]
 * Output: 9
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */
public class ex1_FindMaxElement {

    public int findMax(int[] arr) {
        // TODO: Implement this method
        // Edge case: handle empty or null array
        // Hint: Keep track of the maximum value seen so far while iterating
        if(arr == null || arr.length == 0)
            throw new RuntimeException("array is null or empty");
        int max = arr[0];
        for (int i = 1; i < arr.length; i++) {
            if (arr[i] > max)
                max = arr[i];
        }
        return max;
    }
}